[C2/ONNX] Add support for GatherRanges operator

SplitInfinity opened the pull request:
**Description**
This commit adds support for the [`GatherRanges` operator (click me!)](https://caffe2.ai/docs/operators-catalogue.html#gatherranges). This operator
is not lowered and this commit includes interpreter and CPU backend
implementations for it.

**Testing**
This commit adds two unit tests (one with 32-bit ranges and another with
64-bit ranges) to `OperatorTest` that check operator correctness and one
unit test to `OnnxImporterTest` to check that the operator can be loaded
from a `.onnxtxt` correctly.

**Fixes**
This commit fixes #2210.

yinghai commented on lib/Backends/Interpreter/InterpreterNodes.cpp:
```diff
@@ -876,6 +876,78 @@ void InterpreterFunction::fwdGatherInst(const glow::GatherInst *I) {
+      // Copy the specified data to outT.
+      std::copy(&dataT->getUnsafePtr()[start * dataElementSize],
+                &dataT->getUnsafePtr()[(start + len) * dataElementSize],
+                &outT->getUnsafePtr()[outP]);
```
Do we have some sort of check on whether this will go out of bound?

nadavrot replied:
The TensorHandle API has bounds checks, but raw pointer access does not.

yinghai replied:
Yeah, so here it looks like raw pinter access. Maybe add some checks?

yinghai commented:
Code looks good to me. My question is how we are going to lower this to hardware or is it doable?

rdzhabarov commented on include/glow/Importer/CommonOperatorLoader.h:
```diff
@@ -857,6 +857,27 @@ class CommonOperatorLoader : public ProtobufLoader {
+
+    unsigned_t maxOutputSize;
+    ASSIGN_VALUE_OR_RETURN_ERR(maxOutputSize,
+                               loadInt(dict.find("maxOutputSize")->second));
```
where is 'maxOutputSize' coming from? I cannot see it in https://caffe2.ai/docs/operators-catalogue.html#gatherranges

SplitInfinity replied:
As @artemrakhov and I discussed in #2210, the shape of `output` can not be determined at compile time without a hint. `maxOutputSize` specifies the largest possible size of `output`, and any nodes that use it must use the value of `lengths` to determine how much of `output` is valid data (since `sum(lengths) == output.size()`).

rdzhabarov commented:
@yinghai 
> My question is how we are going to lower this to hardware or is it doable?

Not directly on MM unit, but if there are general purpose CPUs on chip and depending on # cpus there, it might be possible to run with acceleration.

rdzhabarov commented on tests/unittests/OnnxImporterTest.cpp:
```diff
@@ -807,6 +807,34 @@ TEST(onnx, gather) {
 }
 
+/// Test loading GatherRanges from an ONNX model.
+TEST(onnx, gatherranges) {
```
nit: gatherRanges

rdzhabarov commented on tests/unittests/OnnxImporterTest.cpp:
```diff
@@ -807,6 +807,34 @@ TEST(onnx, gather) {
+                           {&data.getType(), &ranges.getType()}, *F);
+    output = EXIT_ON_ERR(onnxLD.getOutputByName("output"));
+  }
+  EE.compile(CompilationMode::Infer, F);
```
why do we need to compile if not running?
Testing just loader part should be fine here and testing Op itself in the OperatorTest.

rdzhabarov commented on lib/Graph/Nodes.cpp:
```diff
@@ -862,6 +862,32 @@ bool GatherNode::verify() const {
 }
 
+bool GatherRangesNode::verify() const {
+  bool isValid = expectCompareTrue(
```
it would be good to reject op earlier in the pipeline, at the moment of actual loading in the loader.

SplitInfinity commented:
**Changes Since Last Revision**

@yinghai
- added bounds checks to interpreter implementation
- added unit test for Caffe2 importer to prove that it works for Caffe2 proto as well (overkill, maybe?)
 
@rdzhabarov 
- added code to reject op during model loading if inputs have incorrect dimensions
- fixed nits

@jfix71 
- added verification that `Data` and `Output` are the same type in `InstrGen`
- refactored unit test in `OperatorTest` using generic function templated on index type
- fixed nits

**Unresolved Issues**
- as @jfix71 pointed out, we use `size_t` for 64-bit width integer implementations instead of `int64_t`. I think this is best left to a separate PR to fix.

rdzhabarov commented:
> as @jfix71 pointed out, we use size_t for 64-bit width integer implementations instead of int64_t. I think this is best left to a separate PR to fix.

If you do not plan to fix it right away, can you file an issue for this?

SplitInfinity commented:
@jackm321 

I opened #2252 to address (most of) your comments.

SplitInfinity commented on lib/Backends/CPU/LLVMIRGen.cpp:
```diff
@@ -2209,6 +2209,45 @@ void LLVMIRGen::generateLLVMIRForInstr(llvm::IRBuilder<> &builder,
+      F = getFunction("gatherranges32", output->getElementType());
+    }
+    if (!F) {
+      llvm_unreachable("Cannot get function for GatherRanges. "
```
I followed what was being done for the `gather` operator above
